File: Dizzo-Studio-Frontend/scripts/strip-rules.test.mjs
// The laser strip's rules (app/lib/design/document.ts), the same cases as
// the backend's tests/test_laser_strip.py.
//   node --experimental-strip-types --test scripts/strip-rules.test.mjs
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { fitIntoStrips, followStrip, placeInStrip, placeInZone, printZone, settleStrips } from '../app/lib/design/document.ts';

// The mug's engraving zone: 75,15 100×50 mm, a 40 mm strip.
const method = {
  method: 'engrave', zone_x_mm: '75', zone_y_mm: '15', zone_w_mm: '100', zone_h_mm: '50',
  max_width_mm: null, max_height_mm: null, strip_width_mm: '40', min_font_mm: '2', colors_allowed: false, dpi: 600,
};
const area = { key: 'wrap', name: 'O‘rab olish', width_mm: '226', height_mm: '79', methods: [method] };
const zone = { x0: 75, y0: 15, x1: 175, y1: 65 };
const text = (id, x, w) => ({ id, area: 'wrap', method: 'engrave', kind: 'text', x_mm: x, y_mm: 40, w_mm: w, h_mm: 10, rotation: 0 });
const span = (lo, hi) => ({ lo, hi });
const nothing = span(Infinity, -Infinity);

test('the strip stays until the design reaches its edge', () => {
  assert.equal(followStrip(100, span(110, 130), zone, 40), 100);
  assert.equal(followStrip(100, span(100, 140), zone, 40), 100);
  assert.equal(followStrip(100, span(95, 120), zone, 40), 95);
  assert.equal(followStrip(100, span(120, 150), zone, 40), 110);
  assert.equal(followStrip(100, span(60, 70), zone, 40), 75);
  assert.equal(followStrip(100, span(170, 190), zone, 40), 135);
  assert.equal(followStrip(100, nothing, zone, 40), 100);
  assert.equal(followStrip(300, nothing, zone, 40), 135);
  assert.equal(followStrip(100, span(100, 160), zone, 40), 100); // wider than the strip: it stays
  assert.equal(followStrip(160, span(100, 160), zone, 40), 135);
});

test('a stored strip is where it was left, else centred', () => {
  const layers = [text('a', 110, 10), text('b', 130, 10)];
  assert.equal(printZone(area, method, layers, []).x0, 100);
  assert.equal(printZone(area, method, layers, [{ area: 'wrap', method: 'engrave', x_mm: 95 }]).x0, 95);
  assert.equal(printZone(area, method, layers, [{ area: 'wrap', method: 'engrave', x_mm: 60 }]).x0, 75);
  assert.equal(printZone(area, method, layers, [{ area: 'wrap', method: 'engrave', x_mm: 170 }]).x0, 135);
});

test('settling adds, follows and drops strips', () => {
  const doc = { version: 1, links: [], layers: [text('a', 110, 10), text('b', 130, 10)] };
  const added = settleStrips(doc, [area]);
  assert.deepEqual(added.strips, [{ area: 'wrap', method: 'engrave', x_mm: 100 }]);
  assert.equal(settleStrips(added, [area]), added); // nothing to change: the same document

  // Moved inside the strip: it stays; past its edge: it follows.
  const inside = { ...added, layers: [text('a', 106, 10), text('b', 130, 10)] };
  assert.equal(settleStrips(inside, [area]).strips[0].x_mm, 100);
  const past = { ...added, layers: [text('a', 110, 10), text('b', 140, 10)] };
  assert.equal(settleStrips(past, [area]).strips[0].x_mm, 105);

  assert.deepEqual(settleStrips({ ...added, layers: [] }, [area]).strips, []);
  assert.equal(settleStrips(added, []), added); // no shape yet: left alone
});

test('a moved layer keeps the design in one strip and in the zone', () => {
  const others = [text('a', 110, 10)]; // 105–115: the other layer may reach 145 at most
  assert.equal(placeInZone(text('b', 160, 10), method, others).x_mm, 140);
  assert.equal(placeInZone(text('b', 130, 10), method, others).x_mm, 130);
  assert.equal(placeInZone(text('b', 70, 10), method, others).x_mm, 80); // the zone's edge
  assert.equal(placeInZone(text('b', 200, 10), method, []).x_mm, 170); // alone: anywhere in the zone
  // An older design that can't fit is left as it is.
  assert.equal(placeInZone(text('b', 160, 50), method, others).x_mm, 160);
});

test('an arriving layer goes into the strip where it is now', () => {
  const others = [text('a', 110, 10)];
  const strips = [{ area: 'wrap', method: 'engrave', x_mm: 100 }]; // 100–140
  assert.equal(placeInStrip(text('b', 150, 10), area, method, others, strips).x_mm, 135);
  assert.equal(placeInStrip(text('b', 90, 10), area, method, others, strips).x_mm, 105);
  assert.equal(placeInStrip(text('b', 120, 10), area, method, others, strips).x_mm, 120);
  // The area's first layer stays: the strip will be centred on it.
  assert.equal(placeInStrip(text('b', 160, 10), area, method, [], []).x_mm, 160);
  // Converted layers end up in one strip.
  const fitted = fitIntoStrips([text('a', 80, 10), text('b', 170, 10)], [area], []);
  assert.deepEqual(fitted.map(l => l.x_mm), [80, 110]);
});

File: Dizzo-Studio-Mobile/test/editor/strip_rules_test.dart
import 'dart:convert';

import 'package:dizzo/features/editor/domain/design_document.dart';
import 'package:dizzo/features/editor/domain/print_area.dart';
import 'package:dizzo/features/editor/presentation/canvas/layer_transform.dart';
import 'package:dizzo/features/editor/presentation/render/design_fonts.dart';
import 'package:flutter_test/flutter_test.dart';

// The laser strip's rules: the same cases as the web's
// scripts/strip-rules.test.mjs (and the backend's tests/test_laser_strip.py).

// The mug's engraving zone: 75,15 100×50 mm, a 40 mm strip.
final _method = AreaMethod.fromJson({
  'method': 'engrave', 'zone_x_mm': '75', 'zone_y_mm': '15', 'zone_w_mm': '100', 'zone_h_mm': '50',
  'max_width_mm': null, 'max_height_mm': null, 'strip_width_mm': '40', 'min_font_mm': '2', 'colors_allowed': false,
  'dpi': 600,
});
final _area = PrintArea(key: 'wrap', name: 'O‘rab olish', widthMm: 226, heightMm: 79, methods: [_method]);
const _zone = Box(75, 15, 175, 65);

Layer _text(String id, double x, double w) =>
    Layer(id: id, area: 'wrap', method: 'engrave', kind: LayerKind.text, x: x, y: 40, w: w, h: 10);

({double lo, double hi}) _span(double lo, double hi) => (lo: lo, hi: hi);
const _nothing = (lo: double.infinity, hi: double.negativeInfinity);
const _at100 = [PrintStrip(area: 'wrap', method: 'engrave', x: 100)];

void main() {
  setUpAll(() => designFontsEnabled = false);

  test('the strip stays until the design reaches its edge', () {
    expect(followStrip(100, _span(110, 130), _zone, 40), 100);
    expect(followStrip(100, _span(100, 140), _zone, 40), 100);
    expect(followStrip(100, _span(95, 120), _zone, 40), 95);
    expect(followStrip(100, _span(120, 150), _zone, 40), 110);
    expect(followStrip(100, _span(60, 70), _zone, 40), 75);
    expect(followStrip(100, _span(170, 190), _zone, 40), 135);
    expect(followStrip(100, _nothing, _zone, 40), 100);
    expect(followStrip(300, _nothing, _zone, 40), 135);
    expect(followStrip(100, _span(100, 160), _zone, 40), 100); // wider than the strip: it stays
    expect(followStrip(160, _span(100, 160), _zone, 40), 135);
  });

  test('a stored strip is where it was left, else centred', () {
    final layers = [_text('a', 110, 10), _text('b', 130, 10)];
    expect(printZone(_area, _method, layers, const []).x0, 100);
    expect(printZone(_area, _method, layers, const [PrintStrip(area: 'wrap', method: 'engrave', x: 95)]).x0, 95);
    expect(printZone(_area, _method, layers, const [PrintStrip(area: 'wrap', method: 'engrave', x: 60)]).x0, 75);
    expect(printZone(_area, _method, layers, const [PrintStrip(area: 'wrap', method: 'engrave', x: 170)]).x0, 135);
  });

  test('settling adds, follows and drops strips', () {
    final doc = DesignDocument(layers: [_text('a', 110, 10), _text('b', 130, 10)]);
    final added = withStrips(doc, [_area]);
    expect(added.strips, _at100);
    expect(identical(withStrips(added, [_area]), added), isTrue);

    final inside = added.copyWith(layers: [_text('a', 106, 10), _text('b', 130, 10)]);
    expect(withStrips(inside, [_area]).strips.single.x, 100);
    final past = added.copyWith(layers: [_text('a', 110, 10), _text('b', 140, 10)]);
    expect(withStrips(past, [_area]).strips.single.x, 105);

    expect(withStrips(added.copyWith(layers: const []), [_area]).strips, isEmpty);
    expect(identical(withStrips(added, const []), added), isTrue);
    // Written like the web's JSON.stringify.
    expect(jsonEncode(added.toJson()['strips']), '[{"area":"wrap","method":"engrave","x_mm":100}]');
  });

  test('a moved layer keeps the design in one strip and in the zone', () {
    final others = [_text('a', 110, 10)];
    expect(placeInZone(_text('b', 160, 10), _method, others).x, 140);
    expect(placeInZone(_text('b', 130, 10), _method, others).x, 130);
    expect(placeInZone(_text('b', 70, 10), _method, others).x, 80);
    expect(placeInZone(_text('b', 200, 10), _method, const []).x, 170);
    expect(placeInZone(_text('b', 160, 50), _method, others).x, 160);
  });

  test('an arriving layer goes into the strip where it is now', () {
    final others = [_text('a', 110, 10)];
    expect(placeInStrip(_text('b', 150, 10), _area, _method, others, _at100).x, 135);
    expect(placeInStrip(_text('b', 90, 10), _area, _method, others, _at100).x, 105);
    expect(placeInStrip(_text('b', 120, 10), _area, _method, others, _at100).x, 120);
    expect(placeInStrip(_text('b', 160, 10), _area, _method, const [], const []).x, 160);
    final fitted = fitIntoStrips([_text('a', 80, 10), _text('b', 170, 10)], [_area], const []);
    expect(fitted.map((l) => l.x), [80, 110]);
  });

  test('a snap never pushes the strip where the finger alone would not', () {
    // Strip 100..140; the layer 125..135 moved 4 mm right: 129..139 (inside).
    // The zone's edge isn't near, but another layer's edge at 141 is.
    final a = _text('a', 130, 10);
    final b = Layer(id: 'b', area: 'wrap', method: 'engrave', kind: LayerKind.text, x: 136, y: 58, w: 10, h: 4);
    final r = moveLayer(_area, a, 4, 0, [a, b], threshold: 3, strips: _at100).layer;
    expect(r.x, 134);
    // Without the strip rule it would have snapped to 131..141.
    expect(moveLayer(_area, a, 4, 0, [a, b], threshold: 3).layer.x, 136);
  });

  test('an edited text steps its size down rather than grow wider than the strip', () {
    final start = Layer(
      id: 't',
      area: 'wrap',
      method: 'engrave',
      kind: LayerKind.text,
      x: 125,
      y: 40,
      w: 10,
      h: 10,
      text: const TextSource(content: 'A', font: 'Roboto', sizeMm: 10, color: monoColor),
    );
    final long = const TextSource(content: 'Salom dunyo', font: 'Roboto', sizeMm: 10, color: monoColor);
    final next = withText(_area, start, long, [start]);
    expect(layoutText(long).w, greaterThan(40));
    expect(layerBox(next).width, lessThanOrEqualTo(40 + 1e-9));
    expect(next.text!.sizeMm, lessThan(10));
    expect(next.text!.sizeMm, greaterThanOrEqualTo(2));
    expect(next.text!.sizeMm * 10, closeTo((next.text!.sizeMm * 10).roundToDouble(), 1e-9)); // tenths of a mm
    // Never below the minimum: a very long text stops at it.
    final huge = long.copyWith(content: 'x' * 400);
    expect(withText(_area, start, huge, [start]).text!.sizeMm, 2);
  });
}
